File: usb/transfer.c
// Copyright 2013 Google Inc.  All rights reserved.
// Copyright 2016 the gousb Authors.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <libusb-1.0/libusb.h>
#include <stdio.h>
#include <string.h>

void print_xfer(struct libusb_transfer *xfer);
void xfer_callback(void *);

void callback(struct libusb_transfer *xfer) {
	xfer_callback(xfer->user_data);
}

int submit(struct libusb_transfer *xfer) {
	xfer->callback = &callback;
	xfer->status = -1;
	return libusb_submit_transfer(xfer);
}

void print_xfer(struct libusb_transfer *xfer) {
	int i;

	printf("Transfer:\n");
	printf("  dev_handle:   %p\n", xfer->dev_handle);
	printf("  flags:        %08x\n", xfer->flags);
	printf("  endpoint:     %x\n", xfer->endpoint);
	printf("  type:         %x\n", xfer->type);
	printf("  timeout:      %dms\n", xfer->timeout);
	printf("  status:       %x\n", xfer->status);
	printf("  length:       %d (act: %d)\n", xfer->length, xfer->actual_length);
	printf("  callback:     %p\n", xfer->callback);
	printf("  user_data:    %p\n", xfer->user_data);
	printf("  buffer:       %p\n", xfer->buffer);
	printf("  num_iso_pkts: %d\n", xfer->num_iso_packets);
	printf("  packets:\n");
	for (i = 0; i < xfer->num_iso_packets; i++) {
		printf("    [%04d] %d (act: %d) %x\n", i,
			xfer->iso_packet_desc[i].length,
			xfer->iso_packet_desc[i].actual_length,
			xfer->iso_packet_desc[i].status);
	}
}

// extract data from a non-isochronous transfer.
int extract_data(struct libusb_transfer *xfer, void *raw, int max, unsigned char *status) {
	int i;
	unsigned char *in = xfer->buffer;
	unsigned char *out = raw;

    int len = xfer->actual_length;
    if (len > max) {
            len = max;
    }
    memcpy(out, in, len);
   	if (xfer->status != 0 && *status == 0) {
            *status = xfer->status;
   	}
    return len;
}

// extract data from an isochronous transfer. Very similar to extract_data, but
// acquires data from xfer->iso_packet_desc array instead of xfer attributes.
int extract_iso_data(struct libusb_transfer *xfer, void *raw, int max, unsigned char *status) {
	int i;
	int copied = 0;
	unsigned char *in = xfer->buffer;
	unsigned char *out = raw;
	for (i = 0; i < xfer->num_iso_packets; i++) {
		struct libusb_iso_packet_descriptor pkt = xfer->iso_packet_desc[i];

		// Copy the data
		int len = pkt.actual_length;
		if (copied + len > max) {
			len = max - copied;
		}
		memcpy(out, in, len);
		copied += len;

		// Increment offsets
		in += pkt.length;
		out += len;

        if (copied == max) {
                break;
        }

		// Extract first error
		if (pkt.status == 0 || *status != 0) {
			continue;
		}
		*status = pkt.status;
	}
	return copied;
}

File: usb/transfer.go
// Copyright 2016 the gousb Authors.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package usb

import (
	"errors"
	"fmt"
	"runtime"
	"sync"
	"time"
)

type usbTransfer struct {
	// mu protects the transfer state.
	mu sync.Mutex
	// xfer is the allocated libusb_transfer.
	xfer *libusbTransfer
	// buf is the buffer allocated for the transfer. Both buf and xfer.buffer
	// point to the same piece of memory.
	buf []byte
	// done is blocking until the transfer is complete and data and transfer
	// status are available.
	done chan struct{}
	// submitted is true if this transfer was passed to libusb through submit()
	submitted bool
}

// submits the transfer. After submit() the transfer is in flight and is owned by libusb.
// It's not safe to access the contents of the transfer until wait() returns.
// Once wait() returns, it's ok to re-use the same transfer structure by calling submit() again.
func (t *usbTransfer) submit() error {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.submitted {
		return errors.New("transfer was already submitted and is not finished yet.")
	}
	t.done = make(chan struct{})
	if err := libusb.submit(t.xfer, t.done); err != nil {
		return err
	}
	t.submitted = true
	return nil
}

// wait waits for libusb to signal the release of transfer data.
// After wait returns, the transfer contents are safe to access
// via t.buf. The number returned by wait indicates how many bytes
// of the buffer were read or written by libusb, and it can be
// smaller than the length of t.buf.
func (t *usbTransfer) wait() (n int, err error) {
	t.mu.Lock()
	defer t.mu.Unlock()
	if !t.submitted {
		return 0, nil
	}
	select {
	case <-time.After(10 * time.Second):
		return 0, fmt.Errorf("wait timed out after 10s")
	case <-t.done:
	}
	t.submitted = false
	n, status := libusb.data(t.xfer)
	if status != LIBUSB_TRANSFER_COMPLETED {
		return n, status
	}
	return n, err
}

// cancel aborts a submitted transfer. The transfer is cancelled
// asynchronously and the user still needs to wait() to return.
func (t *usbTransfer) cancel() error {
	t.mu.Lock()
	defer t.mu.Unlock()
	if !t.submitted {
		return nil
	}
	err := libusb.cancel(t.xfer)
	if err == ERROR_NOT_FOUND {
		// transfer already completed
		return nil
	}
	return err
}

// free releases the memory allocated for the transfer.
// free should be called only if the transfer is not used by libusb,
// i.e. it should not be called after submit() and before wait() returns.
func (t *usbTransfer) free() error {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.submitted {
		return errors.New("free() cannot be called on a submitted transfer until wait() returns")
	}
	libusb.free(t.xfer)
	t.xfer = nil
	t.buf = nil
	t.done = nil
	return nil
}

// newUSBTransfer allocates a new transfer structure for communication with a
// given device/endpoint, with buf as the underlying transfer buffer.
func newUSBTransfer(dev *libusbDevHandle, ei EndpointInfo, buf []byte, timeout time.Duration) (*usbTransfer, error) {
	var isoPackets int
	var isoPktSize uint32
	tt := ei.TransferType()
	if tt == TRANSFER_TYPE_ISOCHRONOUS {
		isoPktSize = ei.MaxIsoPacket
		if len(buf) < int(isoPktSize) {
			isoPktSize = uint32(len(buf))
		}
		isoPackets = len(buf) / int(isoPktSize)
		debug.Printf("New isochronous transfer - buffer length %d, using %d packets of %d bytes each", len(buf), isoPackets, isoPktSize)
	}

	xfer, err := libusb.alloc(dev, ei.Address, tt, timeout, isoPackets, buf)
	if err != nil {
		return nil, err
	}

	if tt == TRANSFER_TYPE_ISOCHRONOUS {
		libusb.setIsoPacketLengths(xfer, isoPktSize)
	}

	t := &usbTransfer{
		xfer: xfer,
		buf:  buf,
	}
	runtime.SetFinalizer(t, func(t *usbTransfer) {
		t.cancel()
		t.wait()
		t.free()
	})
	return t, nil
}
